readthedocs/subscriptions/products.py

```diff
@@ -0,0 +1,120 @@
+        if self.unlimited:
+            return f"{default_description} (unlimited)"
+        if self.value:
+            return f"{default_description} ({self.value})"
+        return default_description
+
+    def to_item(self):
+        """
+        Return a tuple with the feature type and the feature itself.
+
+        Useful to use it as a dictionary item.
+        """
+        return self.type, self
+
+
+@dataclass(slots=True)
+class RTDProduct:
+
+    """A local representation of a Stripe product."""
+
+    stripe_id: str
+    features: dict[str, RTDProductFeature]
+    # If this product should be available to users to purchase.
+    listed: bool = False
+
+    def to_item(self):
+        """
+        Return a tuple with the stripe_id and the product itself.
+
+        Useful to use it as a dictionary item.
+        """
+        return self.stripe_id, self
+
+
+def get_product(stripe_id) -> RTDProduct:
+    """Return the product with the given stripe_id."""
+    return settings.RTD_PRODUCTS.get(stripe_id)
+
+
+def get_listed_products():
+    """Return a list of products that are available to users to purchase."""
+    return [product for product in settings.RTD_PRODUCTS.values() if product.listed]
+
+
+def get_products_with_feature(feature_type):
+    """Return a list of products that have the given feature."""
+    return [
+        product
+        for product in settings.RTD_PRODUCTS.values()
+        if feature_type in product.features
+    ]
+
+
+def get_feature(obj, feature_type) -> RTDProductFeature:
+    """
+    Get the feature object for the given type of the object.
+
+    If the object doesn't have the feature, return the default feature or None.
+
+    :param obj: An organization or project instance.
+    :param type: The type of the feature (readthedocs.subscriptions.constants.TYPE_*).
+    """
+    # Hit the DB only if subscriptions are enabled.
+    if settings.RTD_PRODUCTS:
+        from djstripe import models as djstripe
+
+        from readthedocs.organizations.models import Organization
+        from readthedocs.projects.models import Project
+
+        if isinstance(obj, Project):
+            organization = obj.organizations.first()
+        elif isinstance(obj, Organization):
+            organization = obj
+        else:
+            raise TypeError
+
+        # A subscription can have multiple products, but we only want
+        # the product from the organization that has the feature we are looking for.
+        stripe_products_id = [
+            product.stripe_id for product in get_products_with_feature(feature_type)
+        ]
+        stripe_product_id = (
+            djstripe.Product.objects.filter(
+                id__in=stripe_products_id,
+                prices__subscription_items__subscription__rtd_organization=organization,
+            ).values_list("id", flat=True)
+            # TODO: maybe we should merge features from multiple products?
```

Comment:
Can you expand a little more on this? I'm not sure to understand this question.

Reply:
A subscription can have multiple products, (like the normal basic plan + an extra builder). Right now we are just selecting one product, instead of doing that we could:

- Select the first feature found from iterating over all products (or select the one with the larger value)
- Merge the features from all products (basic plan has 2 builders, and the extra builder plan has 1, so we merge the values of all of them, returning a feature with the number `3`)
- Alternative we can just create a custom product that includes all things from the basic plan + the extra builder.

Since we already have this case in one of our subscriptions, we should make a decision how we want to handle this, cc @ericholscher

Reply:
Hrm, good point. I'd like to make it easy to inclucde "addons" to an existing plan, if at all possible. This gives us a bit more flexibility in how we're defining products (eg. the extra builder). 

I don't have a strong preference for how it's actually implemented. I think selecting the largest is probably the most obvious, and we'd just need to update the `+1 builder` to be `3 total builders` basically? But I'm not 100% sure.